ambry-protocol/src/main/java/com.github.ambry.protocol/MessageInfoAndMetdataListSerde.java

```diff
@@ -0,0 +1,223 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.protocol;
+
+import com.github.ambry.account.Account;
+import com.github.ambry.account.Container;
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.messageformat.MessageMetadata;
+import com.github.ambry.store.MessageInfo;
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.ListIterator;
+
+
+/**
+ * A serde for serializing and deserializing list of message info
+ */
+class MessageInfoAndMetadataListSerde {
+
+  private final List<MessageInfo> messageInfoList;
+  private final List<MessageMetadata> messageMetadataList;
+  static final short VERSION_1 = 1;
+  static final short VERSION_2 = 2;
+  static final short VERSION_3 = 3;
+  static final short VERSION_4 = 4;
+
+  private final short version;
+
+  private static final byte CRC_PRESENT = (byte) 1;
+  private static final byte DELETED = (byte) 1;
+
+  MessageInfoAndMetadataListSerde(List<MessageInfo> messageInfoList, List<MessageMetadata> messageMetadataList,
+      short version) {
+    if (messageMetadataList != null && messageInfoList.size() != messageMetadataList.size()) {
```

Comment:
why can't we check for size match only if messageMetadataList is not Empty. This also avoids NCopies in the next constructor

Reply:
`messageMetadataList` is supposed to be either null or the same size as `messageInfoList`.